doc/ERROR_CONCEPT.md

```diff
@@ -0,0 +1,250 @@
+Alternatively there could be a command for users to see the mountpoint, eg. `kdb get-mountpoint <key>`
+
+### Line 7: At
+
+Reasons to keep it:
+* Debug purposes
+
+Reasons against it:
+* Another extra line in the error message.
+* I highly doubt that most common users will require this information (unnecessary)
+
+Suggestions:
+Remove it for standard error messages. Maybe a `kdb set -v` could include such information for developers, users.
+Alternatively, another possibility would be to have a command which reveals the last error message with more detail: 
+`kdb last-message`.
+
+### Line 6: Module
+
+Reasons to keep it:
+* See the concrete plugin which causes the failure
+* Prohibits error masking (eg. `stat`ing a file happens in more plugins, devs will know where the error occurs)
+
+Reasons against it:
+* Another extra line in the error message.
+
+Suggestions:
+The need for plugin distinguishment in the error message should not be that high in my opinion. Maybe it would make sense to include it in line #1.
+
+### Line 5: Ingroup
+
+Reasons to keep it:
+* Another higher layer of distinguishment for the error location
+
+Reasons against it:
+* Extra line & unnecessary
+
+Suggestion:
+Having both ingroup and module is too much information. Also again in most cases users should get the relevant information out of the reason line.
+As of now there are 6 different "ingroup"s. I would suggest to remove it.
+
+### Line 4: Reason
+
+The most relevant line out of all, no question to keep it or not.
+Still I would remove the prefix `Reason: ` because of the reduced error message length in the new format. It makes it clear
+that this line is the reason. Probably we could substitute it with the plugin which called the error: 
+`<plugin>: Could not set ...`
+
+### Line 3: Description
+
+Reasons to keep it:
+* Gives users a quicker overview about the error (eg. permission error/ resource error/ etc. or another type of categorization such as Markus
+suggested with temporary/permanent/conflict/etc.)
+* Can indicate the type of solution a user has to approach
+
+Reasons against it:
+* Extra line
+* What to do with "uncategorizable" errors
+* Maintenance effort: introduce a new category once there are enough similar errors and change all old errors to this category?
+* Awareness of developers: All devs have to know all categories and choose the correct one. Devs should also be prohibited to throw their error
+into the "uncategorized" category due to lazyness.
+
+Suggestion:
+This line is by far the most discussable one. I thought about categorizations such as
+* Permission/ Resource/ Parse/ External(C++, Ruby, Java)/.../ Miscellaneous:
+Most errors are resource errors (could not stat, no connection, could not close/open/find etc.) This again would not give users
+a particularly useful error indication since he/she has to lookup in the reason anyway. Even subdividing the `Resource` category
+would lead to the same problem. No matter how I look at it, the user
```

Comment:
I think only one division *might* be interesting for a user: Is it something that might get fixed by itself (connection problems) or something that a sysadmin can fix (permission, not enough space, ..)?

But even this division is shaky: some connection problems need to be fixed by admins, or some cron job might remove files (and thus "not enough space" might fix itself)

Reply:
I agree on your division but got some practical concerns. Beside some connection/timeout problem I do not see a lot of temporary issues (especially when looking through the specification file which errors could be temporary). And even in theses cases, usually there is a reason behind it that need to be fixed by sysadmins (proxy settings/firewall settings/power outage of router/etc.) In by far the most cases I think the sysadmin has to fix something. For those cases in which it is a real temporary problem, the user will retry it anyway I think

Reply:
Yes, it is okay for me that we do not distinguish between temporary and permanent errors. Only the conflict, resource and validation errors need to be definitely different (as different reactions are required on these errors).